### PR TITLE
PF-436: Code reorganization. Remove enable/stop pattern.

### PR DESCRIPTION
**Pull request by marikomedlock**
- Changed packaged structure to group by functional area.
- The `terra app enable nextflow` command was just creating a new sub-directory in the workspace directory. I moved this to be a part of the run command. We may still want to do something pre-run in the future (e.g. start a Tower server), but there are still some uncertainties that may affect how it's implemented in the CLI (e.g. 1 per user, 1 per workspace, 1 per Terra).
- Removed the enable/stop pattern. It was hard to find the right way to codify this pattern for apps without any examples. I'll plan to add it back once we have some use cases. I think that would help to decide the tradeoffs and correct structure better. Now it seems too early (i.e. too few apps) to decide on a template for generalizing new additions.
- Renamed TerraUser class' terraUserName property -> terraUserEmail, so it's clearer that this maps to the SAM userEmail field.

### CODE REVIEW

File: src/main/java/bio/terra/cli/context/TerraUser.java
Comment (gjuggler) [inline src/main/java/bio/terra/cli/context/TerraUser.java:1]: This is a tough one – I could see it fitting here in the `context` package, or over in `auth`. Arguably, `auth` may be ever so slightly better as it would reduce the number of cross-package dependencies (since TerraUser depends on AuthUtils, and AuthManager depends on TerraUser).

File: src/main/java/bio/terra/cli/service/utils/WorkspaceManagerUtils.java
Comment (gjuggler) [inline src/main/java/bio/terra/cli/service/utils/WorkspaceManagerUtils.java:1]: Side note: gut feeling, I think we're eventually going to want to rearrange these static methods once we start unit testing more of the codebase. Static methods are notoriously tricky to deal with in those scenarios.

That said, I wouldn't want to prematurely optimize anything yet — especially since this codebase doesn't use our usual Spring dependency injection approach, so my usual aesthetic won't necessarily apply here.

**Review by gjuggler — APPROVED** [review]
Thanks for taking a stab at this! To my eye, this feels like a reasonable improvement. Best to get a second opinion though.

My general feeling is that lots will necessarily be changing in this repo in the next 1-2 sprints, as we build out commands, incorporate feedback from product & solutions folks, and build out test infrastructure. So I think erring on the side of "be willing to quickly iterate towards a stable & predictable code & package organization" will benefit us here.

File: src/main/java/bio/terra/cli/service/utils/WorkspaceManagerUtils.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/service/utils/WorkspaceManagerUtils.java:1]: Actually, I was hoping this would make testing a little easier since we can unit test this class separately -- i.e. test communication with WSM service separately from managing workspace context + communication with WSM service.

Originally I had all these methods as helpers in the WorkspaceManager class, but thought it was slightly more readable to have the WSM service communication separate from the context manipulation. Not sure if this will hold up as we add more to the codebase.

File: src/main/java/bio/terra/cli/context/TerraUser.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/context/TerraUser.java:1]: I prefer having all the classes that define how we persist context on disk in the same place. So other packages look to one place to see the state available to them. If we end up with a lot more context-related classes, then I might go the other way -- distributing them among the other functional packages. I think all the other packages (except server-related commands) will depend on TerraUser, since most of them get the current user as part of their operations, so having auth as the one exception doesn't seem like a large benefit to me.

File: src/main/java/bio/terra/cli/service/utils/WorkspaceManagerUtils.java
Comment (wchamber) [reply src/main/java/bio/terra/cli/service/utils/WorkspaceManagerUtils.java:1]: I think we could make these non-static while still keeping this distinct from the manager class. I'd also tend towards wrapping data and behavior in class instances instead of static methods. It seems like `ApiClient` is the implicit argument always, but do we even need to expose it?

```
/** Calls Workspace Manager endpoints. */
public class WorkspaceManagerService {
  private final ApiClient;

  public WorkspaceManagerService(TerraUser terraUser, ServerSpecification server);
 
  public SystemVersion getVersion();
  WorkspaceDescription createWorkspace();
  ....
}
```

Comment (marikomedlock) [reply src/main/java/bio/terra/cli/service/utils/WorkspaceManagerUtils.java:1]: Okay, I don't have strong feelings about static/not. I mostly want to separate out the code that communicates with Terra service from the code that manipulates the CLI context. I'll refactor these service utils classes to not have static methods.

File: src/main/java/bio/terra/cli/apps/DockerAppsRunner.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/apps/DockerAppsRunner.java:275]: Sure, I'll change to lists.

File: src/main/java/bio/terra/cli/apps/NextflowRunner.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/apps/NextflowRunner.java]: setContext was a method instead of part of the constructor so that it could be specified as part of the interface. Without the interfaces, agreed it works better in a constructor.

I think your comment about how to reflect what current state is required may still apply.
- I thought of the login flow as kind of user-facing input/output, and opted to call AuthenticationManager.loginUser (which launches the oauth flow) in the command classes.
- Many operations require a current user or workspace. I thought of this as a basic context check, and opted to call GlobalContext.requireCurrentTerraUser and WorkspaceContext.requireCurrentWorkspace in the methods where the relevant value will actually be used. This was more for debugging though -- it doesn't really help in terms of forcing you to call loginUser at the command level.

Anyway, that's what I was thinking when I set it up this way. Happy to continue discussion with potentially a follow-up PR to implement any improvements.

File: src/main/java/bio/terra/cli/apps/interfaces/Enable.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/apps/interfaces/Enable.java]: It was the getter for the implementing class. I don't think it's relevant now that I've removed these interfaces.

File: src/main/java/bio/terra/cli/apps/interfaces/Supported.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/apps/interfaces/Supported.java]: I think this was just a copy/paste error -- classes that extended this interface showed up in the "list" command, not "enable".

File: src/main/java/bio/terra/cli/apps/interfaces/Supported.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/apps/interfaces/Supported.java]: That was the Supported interface. List interface seemed like a confusing name, since it didn't have anything to do with a Java list. Supported is also the word used in the user-facing output around `terra app list`.

File: src/main/java/bio/terra/cli/command/app/Enable.java
Comment (marikomedlock) [reply src/main/java/bio/terra/cli/command/app/Enable.java]: I think we don't have enough examples to make good decisions here. We don't even know whether this enable/stop pattern will carry over to other apps. So, I'd like to just postpone this conversation until we have more firm constraints. Thank you for the suggestions though, I'll keep them in mind if we do decide to move forward with this pattern.

Regarding the value of listing supported apps:
There are some "apps"=client-side tools that I think we'd like to call in a pass-through manner. For example:
`terra nextflow run hello`
`terra gcloud ls`
The goal is to have the commands look exactly as they would if you ran them on your local machine, except with a `terra` prefix. The purpose of the prefix is to run the command in the context of a workspace. Currently, this context just means running with the credentials and google project set for you already (see terra_init.sh script). I think it will also soon mean that data references scoped to the workspace are resolved for you already (whatever that means -- env vars, flat file, something more fancy).

These pass-through apps are not visible in the top-level help (that was confusing because they don't really fit into the other command groupings we have there -- e.g. notebooks, workspace, auth). So `terra app list` is a way to see the ones we support.

**Review by wchamber — APPROVED** [review]

File: src/main/java/bio/terra/cli/apps/NextflowRunner.java
Comment (wchamber) [reply src/main/java/bio/terra/cli/apps/NextflowRunner.java]: Sounds good. I think we want to think more about forcing the right setup at the command level, but not now or in this PR.

**Review by gjuggler — APPROVED** [review]

**Pull request merged**
